api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2022.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// All fields in this package are required unless Explicitly marked optional
+// +kubebuilder:validation:Required
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// AdminNetworkPolicy is the Schema for the adminnetworkpolicies API
+type AdminNetworkPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of AdminNetworkPolicy.
+	Spec AdminNetworkPolicySpec `json:"spec,omitempty"`
+
+	// ANPStatus is the status to be reported by the implementation, this is not
+	// standardized in alpha and consumers should report what they see fit in
+	// relation to their AdminNetworkPolicy implementation
+	// +optional
+	Status AdminNetworkPolicyStatus `json:"status,omitempty"`
+}
+
+// AdminNetworkPolicyStatus defines the observed state of AdminNetworkPolicy
+type AdminNetworkPolicyStatus struct {
+	Conditions []metav1.Condition `json:"status"`
+}
+
+// AdminNetworkPolicySubject defines what objects the policy selects.
+// Exactly one of the `NamespaceSelector` or `NamespaceAndPodSelector` pointers
+// should be set.
+type AdminNetworkPolicySubject struct {
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+	// +optional
+	NamespaceAndPodSelector *NamespacedPodSubject `json:"namespaceAndPodSelector"`
```

Comment:
Maybe service selector is less common. Think service account selector -- e.g. most pods can't egress to something, but if you run as some service account you can talk to specific services. In the future if we have multi-cluster semantics, we can easily add cluster selectors here with minimal tweaking of the API.

I don't think we should walk back on the extensibility since it's been such a pain for NetPol.

Reply:
> A user trying to use the service selector would naturally try to do this:
> 
> ```
> "subject": {
>     "namespaceSelector": { ... },
>     "serviceSelector": { ... },
> }
> ```
> 
> Howeverr, an old CNI which is using API version v1 doesn't know about the `ServiceSelector`, so when if parses the spec it sees:
> 
> ```
> "subject": {
>     "namespaceSelector": { ... },
>     "podSelector": nil,
> }
> ```

as discussed in https://github.com/kubernetes-sigs/network-policy-api/pull/30#discussion_r848467838, I think we can prevent plugins from making that mistake

Reply:
Updated the naming of these fields for both the Subject and Peer structs.